### DISCUSSION TITLE
Simplification of empty catch clauses

### DISCUSSION THREAD

**Discussion by softlion** [dotnet/csharplang#5581] [Language Ideas]
try catch is quite verbose with its braces.

```csharp
try
{
    await Task.Delay(timerPeriod, ctsTimerEnabled.Token);
}
catch (TaskCanceledException) 
{
}
```

Instead it could be simpler when the catch clause is empty:

```csharp
try
{
    await Task.Delay(timerPeriod, ctsTimerEnabled.Token);
}
catch (TaskCanceledException);
```

The best of course would be 0 braces:

```csharp
try await Task.Delay(timerPeriod, ctsTimerEnabled.Token);
catch (TaskCanceledException);
```

This verbosity is a reason why devs don't use often a cancellation token with Task.Delay.

A second idea would be to make the "Exception" postfix optional like this in catch clauses. Same as for Attributes.

```csharp
try await Task.Delay(timerPeriod, ctsTimerEnabled.Token);
catch (TaskCanceled);
```

Or make it one line without the middle ";":
```csharp
try await Task.Delay(timerPeriod, ctsTimerEnabled.Token) catch (TaskCanceled);
```

**Comment by bernd5** [comment]
Catching all exception without handling them is in general a very bad idea. It often shadows broken code.
This should not be supported by the language...

**Reply by softlion** [reply]
Where do you see a catch all there ? it's a TaskCanceledException.

**Reply by theunrepentantgeek** [reply]
I suspect a dialectical difference in the English language is at play here. 

If I'm right, this could be rephrased as

> Catching **any** exceptions without handling them is in general a very bad idea. It often shadows broken code.

I agree. I've _rarely_ had exactly nothing to do when catching a `TaskCanceledException`.

**Comment by HaloFour** [comment]
There's nothing that requires you to put each brace on it's own line.  You can simply write this:

```cs
try
{
    await Task.Delay(timerPeriod, ctsTimerEnabled.Token);
}
catch (TaskCanceledException) { }
```

In which case your proposal saves exactly one character.

Empty catch blocks are also considered bad practice, an indicator that the developer does not know how to correctly handle the given exception, and should not be encouraged by the language.

**Reply by softlion** [reply]
Where do you see a catch all there ? it's a TaskCanceledException. Also, you are on 4 lines. I'm on 1.

**Comment by CyrusNajmabadi** [comment]
> Instead it could be simpler when the catch clause is empty:

This saves exactly one character over `catch (Whatever){}`.  So there's practically no way we're going to do this.

Furthermore, for the Task.Delay + cancellation case, you can create a non-throwing awaitable which can allow you to await without needing the try/catch.  For example:

https://github.com/microsoft/vs-threading/blob/main/src/Microsoft.VisualStudio.Threading/TplExtensions.cs#L239

**Reply by softlion** [reply]
The idea of this github is to propose. You seem to like braces a lot. Do you see a brace in yaml or python ? It's time to move forward.

**Reply by jnm2** [reply]
Pointing out that `{}` versus `;` is one character does not equate to "liking braces a lot." YAML and Python are whitespace-sensitive languages, and there's no sense in which whitespace sensitivity is considered 'forward' for C# by the language design team that I've picked up on. A proposal to make semicolons optional was emphatically rejected, in spite of YAML and Python not requiring semicolons.

**Reply by theunrepentantgeek** [reply]
> it is time to move forward

Extremely easy to do something different when you're starting a new language from scratch; nigh on impossible when there's 20+ years of existing code to accomodate. 

As has been discussed in this repo **_many_** times, removing the braces from `try..catch` and `try..finally` would introduce a whole heap of complexity and ambiguity for almost zero return.

> This verbosity is a reason why devs don't use often a cancellation token with Task.Delay.

I don't believe this is a realistic assertion. Any dev who uses language verbosity as an excuse for _not doing their job_ should expect all the appropriate consequences of behaving in such an unprofessional manner.

**Reply by 333fred** [reply]
> Do you see a brace in yaml or python

Both of these languages use whitespace for controlling scope, something C# does not do. I'm not sure what relevance they have to this conversation.

In general, I agree with Cyrus and the other commenters that this proposal isn't worth it's weight. _Every_ feature we add to C# costs, at a minimum, 100 hours of engineer time (and that may be lowballing it). The LDM needs to design and approve of the feature. The compiler team needs to implement the feature and thoroughly test it. The IDE team needs to react to the feature and update features to support the new syntax. Other partner teams (both at Microsoft and 3rd party) need to update their various analyzers or other introspection tools to understand the new syntax. The docs team needs to create documentation for the new feature... That's a _lot_ of work. And overall, we'd be saving one character for a use case that we don't particularly want to encourage. IMO, that doesn't meet the bar.

**Reply by CyrusNajmabadi** [reply]
> I don't believe this is a realistic assertion. 

Furthermore, as i mentioned above.  If you don't want the try/catch, you don't need it.  THere are existing code patterns (and libraries) you can use to write non-throwing awaits of Task.Delay.

**Reply by CyrusNajmabadi** [reply]
> The idea of this github is to propose

We address this here: https://github.com/dotnet/csharplang#discussions

> **Debate** pertaining to language features takes place in the form of Discussions in this repo.

Key word put first (and bolded here).  Everyone is free to make a reasonable proposal here.  But part of the discussion process is that you're going to get *direct* and unfiltered responses from the language designers (and community members) as to how we feel about such proposals.  When i read these I respond giving you exactly the thought process that I put in when evaluating if I think this is something we should move forward on, something we need a better proposal on, or something we think just isn't a fit for the language over any conceivable time period :)
